app/ModelFunctions/LDAPFunctions.php

```diff
@@ -0,0 +1,377 @@
+<?php
+
+namespace App\ModelFunctions;
+
+use App\Exceptions\LDAPException;
+use App\Models\Configs;
+use App\Models\Logs;
+
+class LDAPFunctions
+{
+	/* @var resource $con holds the LDAP connection */
+	protected $con = null;
+
+	/* @var int $bound What type of connection does already exist? */
+	protected $bound = 0; // 0: anonymous, 1: user, 2: superuser
+
+	/* @var array user info queried from LDAP */
+	protected $user_info = [];
+
+	/**
+	 * Log debug message.
+	 *
+	 * @param string $method
+	 * @param string $line
+	 * @param string $text
+	 */
+	protected function _debug(string $method, string $line, string $text)
+	{
+		if ($this->getConf('debug', '0')) {
+			Logs::notice($method, $line, $text);
+		}
+	}
+
+	/**
+	 * Wraps around ldap_search, ldap_list or ldap_read depending on $scope.
+	 *
+	 * @param resource   $link_identifier
+	 * @param string     $base_dn
+	 * @param string     $filter
+	 * @param string     $scope           can be 'base', 'one' or 'sub'
+	 * @param array|null $attributes
+	 * @param int        $attrsonly
+	 * @param int        $sizelimit
+	 *
+	 * @return resource
+	 */
+	protected function _ldapsearch($link_identifier, $base_dn, $filter, $scope = 'sub', $attributes = null,
+						 $attrsonly = 0, $sizelimit = 0)
+	{
+		if (is_null($attributes)) {
+			$attributes = [];
+		}
+
+		if ($scope == 'base') {
+			return @ldap_read(
+				$link_identifier, $base_dn, $filter, $attributes,
+				$attrsonly, $sizelimit
+			);
+		} elseif ($scope == 'one') {
+			return @ldap_list(
+				$link_identifier, $base_dn, $filter, $attributes,
+				$attrsonly, $sizelimit
+			);
+		} else {
+			return @ldap_search(
+				$link_identifier, $base_dn, $filter, $attributes,
+				$attrsonly, $sizelimit
+			);
+		}
+	}
+
+	/**
+	 * Reads the user data from the LDAP server.
+	 *
+	 * @param string $user
+	 *
+	 * @return array containing user data or false
+	 */
+	public function get_user_data($user)
+	{
+		global $conf;
+		if (!$this->_openLDAP()) {
+			return false;
+		}
+		if (!empty($this->user_info) && in_array($user, $this->user_info)) {
+			$this->_debug(__METHOD__, __LINE__, "getUserData: Use cached info for $user");
+
+			return $this->user_info[$user];
+		}
+		// force superuser bind if wanted and not bound as superuser yet
+		if ($this->getConf('binddn') && $this->getConf('bindpw') && $this->bound < 2) {
+			// use superuser credentials
+			if (!@ldap_bind($this->con, $this->getConf('binddn'), $this->getConf('bindpw'))) {
+				Logs::notice(__METHOD__, __LINE__, 'LDAP bind as superuser failed.');
+
+				return false;
+			}
+			$this->bound = 2;
+		}
+
+		$info = [];
+		$info['user'] = $user;
+		$info['server'] = $this->getConf('server');
```

Comment:
Don't use associative array to pass around information or - in this case - as a return type. These kind of arrays are not type-safe and a real PITA, because it is difficult to find out which array keys exist and they are easily mistyped.

With a few remaining exceptions this kind of information transfer has been nuked from Lychee recently.